analyzers/src/SonarAnalyzer.Common/Rules/InsteadOfAnyBase.cs

```diff
@@ -0,0 +1,124 @@
+
+    private ImmutableArray<KnownType> ContainsTypes { get; } = ImmutableArray.Create(
+        KnownType.System_Collections_Generic_HashSet_T,
+        KnownType.System_Collections_Generic_SortedSet_T);
+
+    protected abstract ILanguageFacade<TSyntaxKind> Language { get; }
+
+    public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(existsRule, containsRule);
+
+    protected abstract bool IsSimpleEqualityCheck(TInvocationExpression node, SemanticModel model);
+    protected abstract SyntaxNode GetArgumentExpression(TInvocationExpression invocation, int index);
+    protected abstract bool AreValidOperands(string lambdaVariable, SyntaxNode first, SyntaxNode second);
+
+    protected InsteadOfAnyBase()
+    {
+        existsRule = Language.CreateDescriptor(ExistsDiagnosticId, MessageFormat);
+        containsRule = Language.CreateDescriptor(ContainsDiagnosticId, MessageFormat);
+    }
+
+    protected override void Initialize(SonarAnalysisContext context) =>
+        context.RegisterNodeAction(Language.GeneratedCodeRecognizer, c =>
+        {
+            var invocation = (TInvocationExpression)c.Node;
+
+            if (IsNameEqualTo(invocation, nameof(Enumerable.Any))
+                && Language.Syntax.HasExactlyNArguments(invocation, 1)
+                && Language.Syntax.TryGetOperands(invocation, out var left, out var right)
+                && IsCorrectCall(right, c.SemanticModel)
+                && c.SemanticModel.GetTypeInfo(left).Type is { } type)
+            {
+                if (ExistsTypes.Any(x => type.DerivesFrom(x)))
+                {
+                    RaiseExists(c, invocation);
+                }
+                else if (ContainsTypes.Any(x => type.DerivesFrom(x) && IsSimpleEqualityCheck(invocation, c.SemanticModel)))
+                {
+                    RaiseContains(c, invocation);
+                }
+                else if (type.DerivesFrom(KnownType.System_Collections_Generic_List_T))
+                {
+                    if (IsSimpleEqualityCheck(invocation, c.SemanticModel))
+                    {
+                        RaiseContains(c, invocation);
+                    }
+                    else
+                    {
+                        RaiseExists(c, invocation);
+                    }
+                }
+            }
+        }, Language.SyntaxKind.InvocationExpression);
+
+    protected bool IsNameEqualTo(SyntaxNode node, string name) =>
+        Language.GetName(node).Equals(name, Language.NameComparison);
+
+    protected static bool IsValueTypeOrString(SyntaxNode expression, SemanticModel model) =>
+        model.GetTypeInfo(expression).Type is { } type
+        && (type.IsValueType || type.Is(KnownType.System_String));
+
+    protected bool IsSimpleEqualsInvocation(TInvocationExpression invocation, string lambdaVariableName)
+    {
+        if (IsNameEqualTo(invocation, nameof(Equals)))
```

Comment:
If I am not missing something, you do not know here if this is actually the `Equals` you are looking for.  For example, if a user creates an Equals with two arguments, you will resolve it to the second one.

A more practical and quite common example:

```csharp
var thingy = "foo";
//   HasExactlyNArguments(invocation, 2) is true here
var resolvesToSecondIf = thingy.Equals("bar", StringComparison.CurrentCulture);
```

I think you might want some semantic checks here, because the word "Equals" is very commonplace, so collisions are to be expected.

Reply:
If you can't resolve the symbols without messing with the logic too much, at least make as many FPs as possible in the UTs, just to know that they exist. Althought I think this should be a bit more precise, if possible.

Reply:
I considered having semantic checks but the logic would be messy, as you said. I can add FPs to the tests tho